ARROW-1565: [C++] Implement TopK/BottomK

aocsa opened the pull request:
Heap-based topk can compute these indices in O(n log k) time

lidavidm commented on cpp/src/arrow/compute/api_vector.h:
```diff
@@ -120,6 +120,25 @@ class ARROW_EXPORT SortOptions : public FunctionOptions {
+class ARROW_EXPORT SelectKOptions : public FunctionOptions {
+ public:
+  explicit SelectKOptions(int64_t pivot = 0, std::vector<std::string> keys = {},
+                          std::string keep = "first",
```
We should use an enum instead of a string for modes like this.

lidavidm commented on cpp/src/arrow/compute/api_vector.h:
```diff
@@ -120,6 +120,25 @@ class ARROW_EXPORT SortOptions : public FunctionOptions {
+  int64_t k;
+  std::vector<std::string> keys;
+  std::string keep;
+  SortOrder order;
```
nit: we should have docstrings for these fields.

lidavidm commented on cpp/src/arrow/compute/api_vector.h:
```diff
@@ -120,6 +120,25 @@ class ARROW_EXPORT SortOptions : public FunctionOptions {
   std::vector<SortKey> sort_keys;
 };
 
+/// \brief Partitioning options for NthToIndices
```
nit: this comment should be updated.

lidavidm commented on cpp/src/arrow/compute/api_vector.h:
```diff
@@ -120,6 +120,25 @@ class ARROW_EXPORT SortOptions : public FunctionOptions {
+/// \brief Partitioning options for NthToIndices
+class ARROW_EXPORT SelectKOptions : public FunctionOptions {
+ public:
+  explicit SelectKOptions(int64_t pivot = 0, std::vector<std::string> keys = {},
```
Here it's called pivot instead of k - is that intentional?

lidavidm commented on cpp/src/arrow/compute/api_vector.h:
```diff
@@ -252,6 +271,40 @@ ARROW_EXPORT
+
+/// @TODO
+ARROW_EXPORT
+Result<Datum> BottomK(const Datum& datum, int64_t k, SelectKOptions options,
```
I'm not sure we need all these overloads in C++. They may instead be better suited as static factory methods on the options (SelectKOptions::BottomK and such).

edponce replied:
Also, `keep_duplicates` can be part of the `SelectKOptions`, so that the API is of the form:
```c++
Result<Datum> SelectK(const Datum& datum, SelectKOptions options = SelectKOptions::Defaults(), ExecContext* ctx = NULLPTR);
```

lidavidm commented on cpp/src/arrow/compute/api_vector.cc:
```diff
@@ -111,6 +111,10 @@ static auto kSortOptionsType =
+static auto kSelectKOptionsType = GetFunctionOptionsType<SelectKOptions>(
+    DataMember("k", &SelectKOptions::k), DataMember("keys", &SelectKOptions::keys),
+    DataMember("order", &SelectKOptions::order),
+    DataMember("keep", &SelectKOptions::keep));
```
nit: it'd be good to add a couple instances of the options to the test in function_test.cc: https://github.com/apache/arrow/blob/3da09e51e3f94ca427915c98ac59d72701f728bc/cpp/src/arrow/compute/function_test.cc#L112

lidavidm commented on cpp/src/arrow/testing/gtest_util.h:
```diff
@@ -157,8 +157,7 @@ using NumericArrowTypes =
 using RealArrowTypes = ::testing::Types<FloatType, DoubleType>;
 
-using IntegralArrowTypes = ::testing::Types<UInt8Type, UInt16Type, UInt32Type, UInt64Type,
-                                            Int8Type, Int16Type, Int32Type, Int64Type>;
```
It seems this is WIP but please don't forget to revert this.

lidavidm commented on cpp/src/arrow/util/heap.h:
```diff
@@ -0,0 +1,76 @@
+// A Heap class, is a simple wrapper to make heap operation simpler.
+// This class is immutable by design
+template <typename T, typename Compare = std::less<T>>
+class ARROW_EXPORT Heap {
```
We should have direct tests of the heap itself.

aocsa replied:
Test for Heap and StableHeap were added

lidavidm commented on cpp/src/arrow/util/heap.h:
```diff
@@ -0,0 +1,76 @@
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
```
Can we make this heap_internal.h to make it clear it's not part of our public API?

lidavidm replied:
That way it won't get installed with the rest of the headers, so it'll be usable only within Arrow itself.

lidavidm commented on cpp/src/arrow/util/heap.h:
```diff
@@ -0,0 +1,76 @@
+template <typename T, typename Compare = std::less<T>>
+class ARROW_EXPORT Heap {
+ public:
+  explicit Heap() : values_(), comp_() {}
```
note there are some lints for this file: https://github.com/apache/arrow/pull/11019/checks?check_run_id=3468575773

```
 /arrow/cpp/src/arrow/util/heap.h:33:  Zero-parameter constructors should not be marked explicit.  [runtime/explicit] [5]
/arrow/cpp/src/arrow/util/heap.h:76:  Could not find a newline character at the end of the file.  [whitespace/ending_newline] [5]
```

lidavidm replied:
You can use docker to run linting locally: https://arrow.apache.org/docs/developers/cpp/development.html#code-style-linting-and-ci

lidavidm commented on cpp/src/arrow/util/heap.h:
```diff
@@ -0,0 +1,76 @@
+namespace internal {
+
+// A Heap class, is a simple wrapper to make heap operation simpler.
+// This class is immutable by design
```
The heap is immutable?

aocsa replied:
oh my bad

lidavidm commented on cpp/src/arrow/util/heap.h:
```diff
@@ -0,0 +1,76 @@
+  explicit Heap() : values_(), comp_() {}
+  explicit Heap(const Compare& compare) : values_(), comp_(compare) {}
+
+  Heap(Heap&&) = default;
+  Heap& operator=(Heap&&) = default;
```
You might enjoy ARROW_DEFAULT_MOVE_AND_ASSIGN. https://github.com/apache/arrow/blob/3da09e51e3f94ca427915c98ac59d72701f728bc/cpp/src/arrow/util/macros.h#L34-L36

lidavidm commented on cpp/src/arrow/util/heap.h:
```diff
@@ -0,0 +1,76 @@
+
+  bool Empty() const { return values_.empty(); }
+
+  size_t Size() const { return values_.size(); }
```
nit: for const getter methods like this, they can be named lowercase still. size() is often left as size() and not Size() in our codebase, and top()/empty()/data() could also follow that.

lidavidm commented on cpp/src/arrow/util/heap.h:
```diff
@@ -0,0 +1,76 @@
+    std::push_heap(values_.begin(), values_.end(), comp_);
+  }
+
+  void SetComparator(const Compare& comp) { comp_ = comp; }
```
Is this used?

lidavidm commented on cpp/src/arrow/compute/kernels/select_k_test.cc:
```diff
@@ -0,0 +1,714 @@
+enable_if_t<std::is_same<ArrowType, Time64Type>::value, std::shared_ptr<DataType>>
+TypeToDataType() {
+  return time64(TimeUnit::NANO);
+}
```
You might want to use default_type_instance: https://github.com/apache/arrow/blob/3da09e51e3f94ca427915c98ac59d72701f728bc/cpp/src/arrow/compute/kernels/test_util.h#L123-L144

aocsa replied:
I will wait a little, once `default_type_instance` is part of master I will replace this part of the code.

lidavidm commented on cpp/src/arrow/compute/kernels/select_k_test.cc:
```diff
@@ -0,0 +1,714 @@
+  bool operator()(const Type& lval, const Type& rval) {
+    if (is_floating_type<typename ArrayType::TypeClass>::value) {
+      // NaNs ordered after non-NaNs
+      if (rval != rval) return true;
```
use std::isnan?

aocsa replied:
Not for this case, this functor is used only to test output not to process data. So it is expected that NaN are not part of the output as null-like values were filtered before in TopK/BottomK.

lidavidm replied:
Ok, I think I'm missing something - then why is it comparing rval to itself?

aocsa replied:
Yeap u are right, it is not needed,  I removed it.

lidavidm commented on cpp/src/arrow/compute/kernels/select_k_test.cc:
```diff
@@ -0,0 +1,714 @@
+    if (k < array.length()) {
+      for (uint64_t i = 0; i < (uint64_t)select_k.length(); ++i) {
+        const auto lval = GetLogicalValue(select_k, i);
+        const auto rval = GetLogicalValue(sorted_array, i);
```
Instead of GetLogicalValue and such, you could slice the arrays and use AssertArraysEqual/AssertArraysApproxEqual. With the `verbose == true` parameter this will also print out the mismatched values on failure.

lidavidm replied:
That would also let you get rid of various casts and such.

lidavidm commented on cpp/src/arrow/compute/kernels/select_k_test.cc:
```diff
@@ -0,0 +1,714 @@
+    AssertBottomKArray(ArrayFromJSON(GetType(), values), n);
+  }
+
+  virtual std::shared_ptr<DataType> GetType() = 0;
```
nit: most other such tests in our codebase call this type_singleton

lidavidm commented on cpp/src/arrow/compute/kernels/select_k_test.cc:
```diff
@@ -0,0 +1,714 @@
+  template <SortOrder order>
+  void AssertSelectKArray(const std::shared_ptr<Array> values, int n) {
+    std::shared_ptr<Array> select_k;
+    ASSERT_OK_AND_ASSIGN(select_k, SelectK<order>(*values, n));
```
Why is everything templated instead of passing it as a parameter?

aocsa replied:
for constant parameters I prefer to use templates, hopefully sometime we will be able to use const_if

lidavidm commented on cpp/src/arrow/compute/kernels/select_k_test.cc:
```diff
@@ -0,0 +1,714 @@
+                   const std::vector<std::string>& values, int64_t k,
+                   std::shared_ptr<Array>* out) {
+    ARROW_ASSIGN_OR_RAISE(*out, SelectK<order>(*(ChunkedArrayFromJSON(type, values)), k));
+    PrettyPrint(**out, {}, &std::cerr);
```
Maybe move the ValidateOutput call here so you can't forget it?

lidavidm commented on cpp/src/arrow/compute/kernels/select_k_test.cc:
```diff
@@ -0,0 +1,714 @@
+  std::vector<double> null_probabilities_;
+};
+
+struct TestTopKWithChunkedArray
```
Why not make these checks above in the typed tests?

lidavidm replied:
You can make the 'base' overload of the test helpers take Datum and then use ChunkedArrayFromJSON in the tests themselves, so that you don't have to write too many overloads.

lidavidm replied:
Also, you can probably make one typed test suite that covers all the numeric and most of the temporal types in one go, since the temporal types generally are numeric underneath.

lidavidm commented on cpp/src/arrow/compute/kernels/vector_sort.cc:
```diff
@@ -156,7 +161,7 @@ struct ChunkedArrayResolver {
 
   int64_t num_chunks_;
-  const Array* const* chunks_;
+  const std::vector<const Array*> chunks_;
```
Was there an issue with the original (chunks_ getting invalidated, presumably?)?

aocsa replied:
Yes there was an issue, a holder for this pointer  was needed

lidavidm commented on cpp/src/arrow/compute/kernels/vector_sort.cc:
```diff
@@ -1778,6 +1784,711 @@ class SortIndicesMetaFunction : public MetaFunction {
+const FunctionDoc top_k_doc(
+    "Return the indices that would partition an array array, record batch or table\n"
+    "around a pivot",
+    ("@TODO"), {"input", "k"}, "PartitionNthOptions");
```
I realize these are TODOs but don't forget to update these :)

lidavidm replied:
Is `k` really a parameter? Shouldn't these be unary functions?

aocsa replied:
I though that kernel arity contains the number of required arguments for the function. This function requires at least it is required the input and k.  So arity should be binary.

lidavidm replied:
But `k` is also in the options, so it's being passed twice.

lidavidm reviewed (commented):
You may want to rebase to fix some of the builds.

I think you've mentioned benchmarks, are there results to show/is there benchmark code?

Should SelectK instead be an aggregate function?

lidavidm commented on cpp/src/arrow/compute/kernels/vector_sort.cc:
```diff
@@ -1778,6 +1784,711 @@ class SortIndicesMetaFunction : public MetaFunction {
+    int64_t out_size = static_cast<int64_t>(heap.Size());
+    ARROW_ASSIGN_OR_RAISE(
+        auto take_indices,
+        MakeMutableArrayForFixedSizedType(uint64(), out_size, ctx_->memory_pool()));
```
I wonder if we shouldn't just return the indices, just like the sort kernel.

aocsa replied:
IMO it should return values to match pandas API.

aocsa replied:
Update: I changed these APIs to return only indices like SortIndices.

lidavidm commented on cpp/src/arrow/compute/kernels/vector_sort.cc:
```diff
@@ -1778,6 +1784,711 @@ class SortIndicesMetaFunction : public MetaFunction {
+    }
+    arrow::internal::Heap<uint64_t, decltype(cmp)> heap(cmp);
+    uint64_t* iter = indices_begin;
+    for (; iter != kth_begin && heap.Size() < static_cast<size_t>(k_); ++iter) {
```
Isn't the second condition redundant because of how kth_begin is defined above?

aocsa replied:
it's not redundant for chunked_array as the  heap size will reach at somepoint the capacity k.

lidavidm reviewed (commented):
Sorry, I hit Enter before I meant to. Continuing on…

edponce commented on cpp/src/arrow/compute/api_vector.cc:
```diff
@@ -140,6 +144,15 @@ PartitionNthOptions::PartitionNthOptions(int64_t pivot)
+      keep(keep),
+      order(order) {}
+constexpr char SelectKOptions::kTypeName[];
+
```
The select K algorithm is a general approach to get the topK, bottomK, or median statistic. It seems that SortOrder option is always `Descending` for topK and `Ascending` for bottomK, so I recommend to use an enum for the type of statistic desired instead of specifying ordering. As a user, if I specify `Ascending` it is not intuitive that it corresponds to topK because it depends from which side the sorted data is searched.
```
enum class SelectionOperator {
  TOPK,
  BOTTOMK,
  MEDIAN,  // possibly for another PR
};
```
Also, I am not sure that the`keys` and `keep` options are part of common SelectK algorithms.

lidavidm replied:
`keep` comes from Pandas [nlargest/nsmallest](https://pandas.pydata.org/docs/reference/api/pandas.core.groupby.SeriesGroupBy.nlargest.html) however it is not meaningful here unless there is an ordering defined on the dataset.

`keys` is because you may wish to sort on one column but take values from another. However, as mentioned below, if indices were returned instead of values, you could do this yourself instead of having to build it into the kernel.

edponce replied:
Ok, got it. Thanks for the clarification.

edponce commented on cpp/src/arrow/compute/kernels/vector_sort.cc:
```diff
@@ -1778,6 +1784,711 @@ class SortIndicesMetaFunction : public MetaFunction {
+    }
+    std::function<bool(const HeapItem&, const HeapItem&)> cmp;
+    if (order_ == SortOrder::Ascending) {
+      cmp = [](const HeapItem& left, const HeapItem& right) -> bool {
```
These ordering checks and compare lambdas can be resolved via templates based on the SelectionOperator (topK vs bottomK).

edponce commented on cpp/src/arrow/compute/kernels/vector_sort.cc:
```diff
@@ -1778,6 +1784,711 @@ class SortIndicesMetaFunction : public MetaFunction {
+      *out_cbegin = heap.Top();
+      heap.Pop();
+      --out_cbegin;
+    }
```
The code for extracting top from heap until empty is used in the same manner across the different `SelectKthInternal` methods. Could this be generalized into a common method for `Heap`? Similarly, other duplicated code across methods could be extracted into a single common method.

aocsa replied:
not needed anymore as now it is implemented using `priority_queue`. besides the way it indexes data is not only for uint64_t but also HeapItem for ChunkedArray case.

aocsa commented on cpp/src/arrow/compute/kernels/vector_sort.cc:
```diff
@@ -1778,6 +1784,711 @@ class SortIndicesMetaFunction : public MetaFunction {
+      kth_begin = end_iter;
+    }
+    std::function<bool(uint64_t, uint64_t)> cmp;
+    if (order_ == SortOrder::Ascending) {
```
I see, I will remove these ordering checks by templates based on the SelectionOperator (topK vs bottomK).

aocsa commented:
Thanks @lidavidm, @edponce  for the feedback.  I reduced the code, refactoring it where it was required.   

> You may want to rebase to fix some of the builds.
> 
> I think you've mentioned benchmarks, are there results to show/is there benchmark code?
> 

Here are some benchmark ([code is here](https://gist.github.com/aocsa/3870345fc9b4bcc28372d3b9f8e8f673)) results for [`partition_nth:intro_select` and `partition_nth:heap`,](https://gist.github.com/aocsa/aca0d8c1863576d6652473062b362295) values are int64 scalars with length  from 1 << 20 to 1 << 23 and with 100 of `null_proportion` , `k` is `0,125*length`: 

`partition_nth:intro_select`
```
NthToIndicesInt64/SelectTop1024/32768/10000/min_time:1.000        336418 ns       336337 ns         4134 bytes_per_second=92.9128M/s items_per_second=12.1783M/s null_percent=0.01 size=32.768k
NthToIndicesInt64/SelectTop1024/32768/100/min_time:1.000          431855 ns       431747 ns         3243 bytes_per_second=72.3804M/s items_per_second=9.48705M/s null_percent=1 size=32.768k
NthToIndicesInt64/SelectTop1024/32768/10/min_time:1.000           413660 ns       413553 ns         3380 bytes_per_second=75.5646M/s items_per_second=9.9044M/s null_percent=10 size=32.768k
NthToIndicesInt64/SelectTop1024/32768/2/min_time:1.000            236539 ns       236480 ns         5940 bytes_per_second=132.147M/s items_per_second=17.3207M/s null_percent=50 size=32.768k
NthToIndicesInt64/SelectTop1024/32768/1/min_time:1.000             77369 ns        77349 ns        18102 bytes_per_second=404.011M/s items_per_second=52.9545M/s null_percent=100 size=32.768k
NthToIndicesInt64/SelectTop1024/32768/0/min_time:1.000            336092 ns       336009 ns         4160 bytes_per_second=93.0033M/s items_per_second=12.1901M/s null_percent=0 size=32.768k
NthToIndicesInt64/SelectTop1024/1048576/100/min_time:1.000       7809018 ns      7807367 ns          179 bytes_per_second=128.084M/s items_per_second=16.7882M/s null_percent=1 size=1048.58k
NthToIndicesInt64/SelectTop1024/8388608/100/min_time:1.000      61132755 ns     61121366 ns           23 bytes_per_second=130.887M/s items_per_second=17.1556M/s null_percent=1 size=8.38861M

NthToIndicesInt64/SelectTop64/32768/10000/min_time:1.000          244987 ns       244925 ns         5716 bytes_per_second=127.59M/s items_per_second=16.7235M/s null_percent=0.01 size=32.768k
NthToIndicesInt64/SelectTop64/32768/100/min_time:1.000            367908 ns       367813 ns         3807 bytes_per_second=84.9616M/s items_per_second=11.1361M/s null_percent=1 size=32.768k
NthToIndicesInt64/SelectTop64/32768/10/min_time:1.000             324431 ns       324349 ns         4354 bytes_per_second=96.3469M/s items_per_second=12.6284M/s null_percent=10 size=32.768k
NthToIndicesInt64/SelectTop64/32768/2/min_time:1.000              220053 ns       219995 ns         6377 bytes_per_second=142.049M/s items_per_second=18.6186M/s null_percent=50 size=32.768k
NthToIndicesInt64/SelectTop64/32768/1/min_time:1.000               80327 ns        80304 ns        17677 bytes_per_second=389.148M/s items_per_second=51.0064M/s null_percent=100 size=32.768k
NthToIndicesInt64/SelectTop64/32768/0/min_time:1.000              247086 ns       246991 ns         5561 bytes_per_second=126.523M/s items_per_second=16.5836M/s null_percent=0 size=32.768k
NthToIndicesInt64/SelectTop64/1048576/100/min_time:1.000         7572595 ns      7570022 ns          185 bytes_per_second=132.1M/s items_per_second=17.3146M/s null_percent=1 size=1048.58k
NthToIndicesInt64/SelectTop64/8388608/100/min_time:1.000        60670175 ns     60652316 ns           23 bytes_per_second=131.899M/s items_per_second=17.2883M/s null_percent=1 size=8.38861M
```

`partition_nth:heap`
```


NthToIndicesInt64/HeapBasedTop1024/32768/10000/min_time:1.000     333859 ns       333737 ns         4180 bytes_per_second=93.6365M/s items_per_second=12.2731M/s null_percent=0.01 size=32.768k
NthToIndicesInt64/HeapBasedTop1024/32768/100/min_time:1.000       433951 ns       433841 ns         3261 bytes_per_second=72.031M/s items_per_second=9.44125M/s null_percent=1 size=32.768k
NthToIndicesInt64/HeapBasedTop1024/32768/10/min_time:1.000        414340 ns       414237 ns         3390 bytes_per_second=75.4399M/s items_per_second=9.88806M/s null_percent=10 size=32.768k
NthToIndicesInt64/HeapBasedTop1024/32768/2/min_time:1.000         238940 ns       238870 ns         5861 bytes_per_second=130.824M/s items_per_second=17.1474M/s null_percent=50 size=32.768k
NthToIndicesInt64/HeapBasedTop1024/32768/1/min_time:1.000          79213 ns        79190 ns        17821 bytes_per_second=394.622M/s items_per_second=51.7238M/s null_percent=100 size=32.768k
NthToIndicesInt64/HeapBasedTop1024/32768/0/min_time:1.000         337863 ns       337777 ns         4125 bytes_per_second=92.5167M/s items_per_second=12.1263M/s null_percent=0 size=32.768k
NthToIndicesInt64/HeapBasedTop1024/1048576/100/min_time:1.000    7984756 ns      7982600 ns          177 bytes_per_second=125.272M/s items_per_second=16.4197M/s null_percent=1 size=1048.58k
NthToIndicesInt64/HeapBasedTop1024/8388608/100/min_time:1.000   61850326 ns     61825764 ns           22 bytes_per_second=129.396M/s items_per_second=16.9602M/s null_percent=1 size=8.38861M

NthToIndicesInt64/HeapBasedTop64/32768/10000/min_time:1.000       246324 ns       246264 ns         5648 bytes_per_second=126.896M/s items_per_second=16.6326M/s null_percent=0.01 size=32.768k
NthToIndicesInt64/HeapBasedTop64/32768/100/min_time:1.000         369914 ns       369818 ns         3778 bytes_per_second=84.5011M/s items_per_second=11.0757M/s null_percent=1 size=32.768k
NthToIndicesInt64/HeapBasedTop64/32768/10/min_time:1.000          323143 ns       323059 ns         4286 bytes_per_second=96.7316M/s items_per_second=12.6788M/s null_percent=10 size=32.768k
NthToIndicesInt64/HeapBasedTop64/32768/2/min_time:1.000           229702 ns       229619 ns         6399 bytes_per_second=136.095M/s items_per_second=17.8383M/s null_percent=50 size=32.768k
NthToIndicesInt64/HeapBasedTop64/32768/1/min_time:1.000            80524 ns        80503 ns        17146 bytes_per_second=388.186M/s items_per_second=50.8804M/s null_percent=100 size=32.768k
NthToIndicesInt64/HeapBasedTop64/32768/0/min_time:1.000           248449 ns       248391 ns         5579 bytes_per_second=125.81M/s items_per_second=16.4902M/s null_percent=0 size=32.768k
NthToIndicesInt64/HeapBasedTop64/1048576/100/min_time:1.000      8001923 ns      7998633 ns          181 bytes_per_second=125.021M/s items_per_second=16.3868M/s null_percent=1 size=1048.58k
NthToIndicesInt64/HeapBasedTop64/8388608/100/min_time:1.000     61181748 ns     61162773 ns           23 bytes_per_second=130.799M/s items_per_second=17.144M/s null_percent=1 size=8.38861M

```
Some conclusions: Both implementations (`partition_nth:intro_select` and `partition_nth:heap`) show similar numbers.  PartitionNthIndices uses std::nth_element which uses internally a special algorithm called [introselect](https://github.com/tupipa/sva_freebsd90/blob/640747eb65a5723250fefc612de48694570292d1/contrib/libstdc%2B%2B/include/bits/stl_algo.h#L3945),  this is a  is a selection algorithm that is a hybrid of quickselect and heap selection.  

This PR started  as a way to speed top-k queries, https://lemire.me/blog/2017/06/21/top-speed-for-top-k-queries/ using `FastPriorityQueue-KWillets-replaceTop` approach.  However C++ implementation for std::nth_element  has the best of quickselect and heap selection as an hybrid solution.

In the next  benchmarks I will show the peformance when the algorithm needs to be stable to enable `keep=first` and `keep=last`. I will explore ways to implement a stable binary heap to do that and compare with  {std::nth_element + std::stable_sort}. 

--
<br class="Apple-interchange-newline">

> Should SelectK instead be an aggregate function?

I am not sure, I am new with arrow. And this function looks like pretty similar to the APIs of SortIndices and PartitionNthIndices So I think it is in the right place.  

Looking forward to your thoughts. cc @lidavidm

lidavidm commented on cpp/src/arrow/compute/kernels/vector_sort.cc:
```diff
@@ -1778,6 +1799,736 @@ class SortIndicesMetaFunction : public MetaFunction {
+                               TakeOptions::NoBoundsCheck(), ctx_));
+    ARROW_ASSIGN_OR_RAISE(
+        auto select_k,
+        Concatenate(chunked_select_k.chunked_array()->chunks(), ctx_->memory_pool()));
```
Why do we concatenate here? If the input is a chunked array I would expect the output to also be a chunked array.

aocsa replied:
Update: I changed these APIs to return only indices like SortIndices.

lidavidm commented on cpp/src/arrow/compute/kernels/vector_sort.cc:
```diff
@@ -1778,6 +1799,736 @@ class SortIndicesMetaFunction : public MetaFunction {
+  std::vector<ResolvedSortKey> sort_keys_;
+  Comparator comparator_;
+  Status status_;
+};
```
These selecters all have a relatively similar core structure. It might be worth considering how some templating (e.g. via [CRTP](https://en.wikipedia.org/wiki/Curiously_recurring_template_pattern), or via a set of helper comparator/iteration templates) could let you factor out the core algorithm and the container-specific bits. It would then be easier to also try to share generated code between types with the same physical type (e.g. as mentioned, Int64, Timestamp, and Date64 should all use the same generated code underneath).

lidavidm replied:
Ping here - we should at least file a followup. The selecters are all very similar and I believe could be consolidated somewhat.

aocsa replied:
Sure,  I will create an issue on JIRA for this case as well .

aocsa replied:
https://issues.apache.org/jira/browse/ARROW-13971

lidavidm commented on cpp/src/arrow/compute/kernels/vector_sort.cc:
```diff
@@ -1778,6 +1799,736 @@ class SortIndicesMetaFunction : public MetaFunction {
+  Status status_;
+};
+
+template <SortOrder sort_order>
```
Another thought: instead of a single SortOrder, why not accept SortKeys like the sort kernels? Then you could emulate a query like `... ORDER BY foo ASC, bar DESC LIMIT 5` which this kernel can't currently do.

aocsa replied:
Sure, I create a new kernel called `Result select_k(datum, options)` where `options.sort_keys` type is `std::vector<SortKey>`.

lidavidm commented on cpp/src/arrow/compute/kernels/select_k_test.cc:
```diff
@@ -0,0 +1,809 @@
+  }
+};
+
+TYPED_TEST_SUITE(TestTopKWithChunkedArray, SelectKableNumericAndTemporal);
```
For chunked array tests, you could get it "for free" by slicing a non-chunked array and building a chunked array to test. This is what the scalar kernel tests do. That way, you don't have to duplicate test cases.

lidavidm replied:
You could imagine doing the same thing for tables/record batches.

lidavidm reviewed (commented):
Thanks for the updates.

We may actually want this to be implemented multiple ways: as both a hash aggregate kernel and a vector kernel. We'll want the hash aggregate kernel to match Pandas, where [nlargest](https://pandas.pydata.org/docs/reference/api/pandas.core.groupby.SeriesGroupBy.nlargest.html) is part of a group by operation. Of course, we can split out the implementation of that for later, but I think it's worth considering how we can share the implementations as much as possible. (It may not be that you can share much, since the hash aggregate kernel has to work in a streaming fashion which this does not do.)

We'll still want a vector kernel implementation (the one here) as essentially a pipeline sink, since this implements an `... ORDER BY ... LIMIT` operation. (We'll need to implement an ExecNode for that - that can be a separate follow up JIRA, it should be straightforward.) Note if you do want to match the API of SortIndices this should return indices, not values.

Also, note there are some CI failures - mostly just different compilers looking for different things.

It might be good to compare the kernel here with SortIndices (since you could emulate a top-K by calling SortIndices, slicing the output, then calling Take) - I wonder how much of a speedup this achieves.

aocsa commented:
Thanks for the feedback @lidavidm 

> Thanks for the updates.
> 
> We may actually want this to be implemented multiple ways: as both a hash aggregate kernel and a vector kernel. We'll want the hash aggregate kernel to match Pandas, where [nlargest](https://pandas.pydata.org/docs/reference/api/pandas.core.groupby.SeriesGroupBy.nlargest.html) is part of a group by operation. Of course, we can split out the implementation of that for later, but I think it's worth considering how we can share the implementations as much as possible. (It may not be that you can share much, since the hash aggregate kernel has to work in a streaming fashion which this does not do.)
> 
> We'll still want a vector kernel implementation (the one here) as essentially a pipeline sink, since this implements an `... ORDER BY ... LIMIT` operation. (We'll need to implement an ExecNode for that - that can be a separate follow up JIRA, it should be straightforward.) Note if you do want to match the API of SortIndices this should return indices, not values.
> 

Update: I changed these APIs to return only indices like SortIndices.

> Also, note there are some CI failures - mostly just different compilers looking for different things.
> 
> It might be good to compare the kernel here with SortIndices (since you could emulate a top-K by calling SortIndices, slicing the output, then calling Take) - I wonder how much of a speedup this achieves.

Here the benchmarks results for  `topk with sortingIndices` and `topk with NonStableHeap`, values are int64 scalars with length from 1 << 20 to 1 << 23 and with 100 of null_proportion , k is 1/8*N:


`TopK_SortingIndices`
```
-----------------------------------------------------------------------------------------------
Benchmark                                     Time             CPU   Iterations UserCounters...
-----------------------------------------------------------------------------------------------
TopKInt64/32768/10000/min_time:1.000     199633 ns       199601 ns         7010 bytes_per_second=156.563M/s items_per_second=20.521M/s null_percent=0.01 size=32.768k
TopKInt64/32768/100/min_time:1.000       205520 ns       205486 ns         6734 bytes_per_second=152.078M/s items_per_second=19.9332M/s null_percent=1 size=32.768k
TopKInt64/32768/10/min_time:1.000        196022 ns       195979 ns         7124 bytes_per_second=159.456M/s items_per_second=20.9002M/s null_percent=10 size=32.768k
TopKInt64/32768/2/min_time:1.000         124550 ns       124531 ns        11204 bytes_per_second=250.941M/s items_per_second=32.8914M/s null_percent=50 size=32.768k
TopKInt64/32768/1/min_time:1.000           9395 ns         9394 ns       149397 bytes_per_second=3.24871G/s items_per_second=436.034M/s null_percent=100 size=32.768k
TopKInt64/32768/0/min_time:1.000         199116 ns       199086 ns         7016 bytes_per_second=156.968M/s items_per_second=20.5741M/s null_percent=0 size=32.768k
TopKInt64/1048576/100/min_time:1.000   10200499 ns     10199057 ns          137 bytes_per_second=98.0483M/s items_per_second=12.8514M/s null_percent=1 size=1048.58k
TopKInt64/8388608/100/min_time:1.000  105033049 ns    105007534 ns           13 bytes_per_second=76.185M/s items_per_second=9.98572M/s null_percent=1 size=8.38861M
```

`TopK_NonStableHeap`
```
-----------------------------------------------------------------------------------------------
Benchmark                                     Time             CPU   Iterations UserCounters...
-----------------------------------------------------------------------------------------------
TopKInt64/32768/10000/min_time:1.000      74202 ns        74191 ns        18553 bytes_per_second=421.208M/s items_per_second=55.2085M/s null_percent=0.01 size=32.768k
TopKInt64/32768/100/min_time:1.000        78732 ns        78719 ns        17032 bytes_per_second=396.982M/s items_per_second=52.0333M/s null_percent=1 size=32.768k
TopKInt64/32768/10/min_time:1.000         79833 ns        79821 ns        17285 bytes_per_second=391.5M/s items_per_second=51.3147M/s null_percent=10 size=32.768k
TopKInt64/32768/2/min_time:1.000          71248 ns        71237 ns        19374 bytes_per_second=438.674M/s items_per_second=57.4979M/s null_percent=50 size=32.768k
TopKInt64/32768/1/min_time:1.000           4680 ns         4679 ns       301875 bytes_per_second=6.52247G/s items_per_second=875.431M/s null_percent=100 size=32.768k
TopKInt64/32768/0/min_time:1.000          73972 ns        73962 ns        18658 bytes_per_second=422.516M/s items_per_second=55.38M/s null_percent=0 size=32.768k
TopKInt64/1048576/100/min_time:1.000    5714238 ns      5713428 ns          247 bytes_per_second=175.026M/s items_per_second=22.941M/s null_percent=1 size=1048.58k
TopKInt64/8388608/100/min_time:1.000   67560407 ns     67546834 ns           20 bytes_per_second=118.436M/s items_per_second=15.5237M/s null_percent=1 size=8.38861M
```

These result show that `TopK_NonStableHeap` is quite faster (around 2.7x).  However  the implementation of a stable version of it using a StableHeap is not as good as `TopK_NonStableHeap`, as you can  see  below:

`TopK_StableHeap`
```
-----------------------------------------------------------------------------------------------
Benchmark                                     Time             CPU   Iterations UserCounters...
-----------------------------------------------------------------------------------------------
TopKInt64/32768/10000/min_time:1.000     213637 ns       213609 ns         6596 bytes_per_second=146.295M/s items_per_second=19.1752M/s null_percent=0.01 size=32.768k
TopKInt64/32768/100/min_time:1.000       222127 ns       222091 ns         6332 bytes_per_second=140.708M/s items_per_second=18.4429M/s null_percent=1 size=32.768k
TopKInt64/32768/10/min_time:1.000        216600 ns       216562 ns         6353 bytes_per_second=144.3M/s items_per_second=18.9137M/s null_percent=10 size=32.768k
TopKInt64/32768/2/min_time:1.000         186647 ns       186613 ns         7585 bytes_per_second=167.459M/s items_per_second=21.9492M/s null_percent=50 size=32.768k
TopKInt64/32768/1/min_time:1.000           8101 ns         8100 ns       173012 bytes_per_second=3.7677G/s items_per_second=505.692M/s null_percent=100 size=32.768k
TopKInt64/32768/0/min_time:1.000         213979 ns       213950 ns         6595 bytes_per_second=146.062M/s items_per_second=19.1447M/s null_percent=0 size=32.768k
TopKInt64/1048576/100/min_time:1.000   11343912 ns     11341499 ns          122 bytes_per_second=88.1718M/s items_per_second=11.5568M/s null_percent=1 size=1048.58k
TopKInt64/8388608/100/min_time:1.000  123205686 ns    123179459 ns           12 bytes_per_second=64.9459M/s items_per_second=8.51259M/s null_percent=1 size=8.38861M

```

This implementation for stable select_k with a stable heap, although it is valid, is still simple and maybe the best choise here is a sort-based algorithm.  IMO, implementation of stable algorithms need more exploration, and it could be implemented in a separate follow up JIRA issue adding an optional parameter to choose between `SortType{kStable, kNonStable}` implementation. 

Besides these updates, instead of a single SortOrder for RecordBatch and Table  a new kernel called `Result select_k(datum, options)` was created  where `options.sort_keys` type is `std::vector<SortKey<string, order>>`.

Looking forward to your thoughts. cc @lidavidm

pitrou commented:
> Some conclusions: Both implementations (partition_nth:intro_select and partition_nth:heap) show similar numbers. 

Interesting. Did you double-check that no mistake was made when running the benchmarks (i.e. are you really comparing the two different algorithms)?

> IMO, implementation of stable algorithms need more exploration, and it could be implemented in a separate follow up JIRA issue

Hmm, why not, but `sort_indices` and `partition_nth_indices` are guaranteed to be stable, so it would be better if topK/bottomK wasn't inconsistent.

lidavidm commented:
Thanks for the benchmarks. It looks like the stable heap is about the same or perhaps slightly slower than just sorting and slicing, which is unfortunate. 

> Hmm, why not, but sort_indices and partition_nth_indices are guaranteed to be stable, so it would be better if topK/bottomK wasn't inconsistent.

Since the unstable version is quite a bit faster so far here, maybe we should expose a `select_k_unstable`? (Was the reason why we don't have an unstable sort because it's not useful and/or wasn't meaningfully faster?)

pitrou commented:
Le 07/09/2021 à 15:46, David Li a écrit :
>     Hmm, why not, but sort_indices and partition_nth_indices are
>     guaranteed to be stable, so it would be better if topK/bottomK
>     wasn't inconsistent.
> 
> Since the unstable version is quite a bit faster so far here, maybe we 
> should expose a |select_k_unstable|?

If "quite a bit" means 2x or 3x rather than 20%, then that sounds good 
to me indeed.

 > (Was the reason why we don't have
> an unstable sort because it's not useful and/or wasn't meaningfully faster?)

I'm not sure we have actually tried to benchmark an unstable sort.  But 
being stable is usually a desirable guarantee when the sort columns 
don't span all columns.  Of course, that depends on the use case :-)

pitrou commented:
Another thought: isn't one attraction of TopK to have a streaming algorithm with O(k) memory consumption? Making a full sort requires to materialize the entire input, hence O(n) memory consumption.

lidavidm commented:
> Another thought: isn't one attraction of TopK to have a streaming algorithm with O(k) memory consumption? Making a full sort requires to materialize the entire input, hence O(n) memory consumption.

That is a fair point. However, this kernel as-is is not a streaming implementation and we would have to implement it as an actual aggregate kernel to get that behavior.

aocsa commented:
> > Another thought: isn't one attraction of TopK to have a streaming algorithm with O(k) memory consumption? Making a full sort requires to materialize the entire input, hence O(n) memory consumption.
> 
> That is a fair point. However, this kernel as-is is not a streaming implementation and we would have to implement it as an actual aggregate kernel to get that behavior.

I have a question related to that, the streaming implementation would be using ExecNodes right?

aocsa commented:
> > Some conclusions: Both implementations (partition_nth:intro_select and partition_nth:heap) show similar numbers.
> 
> Interesting. Did you double-check that no mistake was made when running the benchmarks (i.e. are you really comparing the two different algorithms)?
> 

My bad, You were right  @pitrou  the numbers are not correct.   Something happenned when I ran this [bench](https://gist.github.com/aocsa/3870345fc9b4bcc28372d3b9f8e8f673) both at the same time.  I ran again PartitionNthToIndices with different strategies (nth_element vs heap_based) separetly and I got these numbers: 

`PartitionNthToIndices_nth_element`
```
--------------------------------------------------------------------------------------------------
Benchmark                                        Time             CPU   Iterations UserCounters...
--------------------------------------------------------------------------------------------------
BottomKInt64/32768/10000/min_time:1.000       8558 ns         8556 ns       166021 bytes_per_second=3.56686G/s items_per_second=478.735M/s null_percent=0.01 size=32.768k
BottomKInt64/32768/100/min_time:1.000        20545 ns        20542 ns        63571 bytes_per_second=1.4856G/s items_per_second=199.394M/s null_percent=1 size=32.768k
BottomKInt64/32768/10/min_time:1.000         17864 ns        17862 ns        72315 bytes_per_second=1.70853G/s items_per_second=229.316M/s null_percent=10 size=32.768k
BottomKInt64/32768/2/min_time:1.000          11565 ns        11564 ns       118026 bytes_per_second=2.63912G/s items_per_second=354.217M/s null_percent=50 size=32.768k
BottomKInt64/32768/1/min_time:1.000           6317 ns         6316 ns       214862 bytes_per_second=4.83149G/s items_per_second=648.472M/s null_percent=100 size=32.768k
BottomKInt64/32768/0/min_time:1.000           8407 ns         8406 ns       164849 bytes_per_second=3.63065G/s items_per_second=487.297M/s null_percent=0 size=32.768k
BottomKInt64/1048576/100/min_time:1.000     916082 ns       915966 ns         1526 bytes_per_second=1091.74M/s items_per_second=143.097M/s null_percent=1 size=1048.58k
BottomKInt64/8388608/100/min_time:1.000    7874118 ns      7871600 ns          178 bytes_per_second=1016.31M/s items_per_second=133.21M/s null_percent=1 size=8.38861M
``` 
`PartitionNthToIndices_heap_based`

```
--------------------------------------------------------------------------------------------------
Benchmark                                        Time             CPU   Iterations UserCounters...
--------------------------------------------------------------------------------------------------
BottomKInt64/32768/10000/min_time:1.000      55314 ns        55303 ns        25080 bytes_per_second=565.064M/s items_per_second=74.0641M/s null_percent=0.01 size=32.768k
BottomKInt64/32768/100/min_time:1.000        57577 ns        57568 ns        23965 bytes_per_second=542.835M/s items_per_second=71.1505M/s null_percent=1 size=32.768k
BottomKInt64/32768/10/min_time:1.000         59858 ns        59847 ns        22977 bytes_per_second=522.168M/s items_per_second=68.4416M/s null_percent=10 size=32.768k
BottomKInt64/32768/2/min_time:1.000          54008 ns        53995 ns        24750 bytes_per_second=578.753M/s items_per_second=75.8583M/s null_percent=50 size=32.768k
BottomKInt64/32768/1/min_time:1.000           5929 ns         5927 ns       235385 bytes_per_second=5.14856G/s items_per_second=691.028M/s null_percent=100 size=32.768k
BottomKInt64/32768/0/min_time:1.000          55751 ns        55741 ns        24636 bytes_per_second=560.626M/s items_per_second=73.4823M/s null_percent=0 size=32.768k
BottomKInt64/1048576/100/min_time:1.000    4542982 ns      4541548 ns          323 bytes_per_second=220.189M/s items_per_second=28.8606M/s null_percent=1 size=1048.58k
BottomKInt64/8388608/100/min_time:1.000   53349325 ns     53333611 ns           26 bytes_per_second=149.999M/s items_per_second=19.6607M/s null_percent=1 size=8.38861M
```
> > IMO, implementation of stable algorithms need more exploration, and it could be implemented in a separate follow up JIRA issue
> 
> Hmm, why not, but `sort_indices` and `partition_nth_indices` are guaranteed to be stable, so it would be better if topK/bottomK wasn't inconsistent.

Note that [both algorithms](https://gist.github.com/aocsa/7a2094159e65c69b5e84c694498f03bd) uses `NonStablePartitioner` to separete Nulls and both are unstable.
Clearly, `PartitionNthToIndices_nth_element` is faster (5-6x) but `PartitionNthToIndices` kernel is not enough to implement `TopK/BottomK` because the first `K` elements needs to be sorted. Besides in order to Implement a stable version most probably we will need to use   `StablePartitioner + std::stable_partition + std::stable_sort`.  However  the attraction of  the heap-based implementation  is to have a streaming algorithm. 

IMO this PR should first  expose a `select_k_unstable` and look in a follow up PR the best implementation for `select_k_stable`. 
Looking forward to your thoughts. cc @lidavidm, @pitrou

edponce commented on cpp/src/arrow/compute/api_vector.h:
```diff
@@ -252,6 +296,53 @@ ARROW_EXPORT
+/// \param[in] options options
+/// \param[in] ctx the function execution context, optional
+/// \return a datum with the same schema as the input
+
```
Remove extra whitespace.

edponce commented on cpp/src/arrow/compute/api_vector.h:
```diff
@@ -252,6 +296,53 @@ ARROW_EXPORT
+/// \param[in] ctx the function execution context, optional
+/// \return a datum with the same schema as the input
+
+ARROW_EXPORT
```
Remove blank line.

edponce commented on cpp/src/arrow/compute/api_vector.h:
```diff
@@ -252,6 +296,53 @@ ARROW_EXPORT
+
+ARROW_EXPORT
+Result<std::shared_ptr<Array>> SelectK(const Datum& datum, SelectKOptions options,
+                                       ExecContext* ctx = NULLPTR);
```
Set the default options to `options`. Similarly, for topK and bottomK.

pitrou commented:
> Note that both algorithms uses NonStablePartitioner to separete Nulls and both are unstable.

Oh, you're right! I had forgotten that `partition_nth_indices` was unstable, unlike `sort_indices`.

> Clearly, PartitionNthToIndices_nth_element is faster (5-6x) but PartitionNthToIndices kernel is not enough to implement TopK/BottomK because the first K elements needs to be sorted. 

But if K is small, sorting them should be cheap, right? I'm not sure what the cutoff would be... (sqrt(N) perhaps?)

> IMO this PR should first expose a select_k_unstable and look in a follow up PR the best implementation for select_k_stable

Agreed!

lidavidm commented:
> > > Another thought: isn't one attraction of TopK to have a streaming algorithm with O(k) memory consumption? Making a full sort requires to materialize the entire input, hence O(n) memory consumption.
> > 
> > 
> > That is a fair point. However, this kernel as-is is not a streaming implementation and we would have to implement it as an actual aggregate kernel to get that behavior.
> 
> I have a question related to that, the streaming implementation would be using ExecNodes right?

It's unclear to me. The API for scalar aggregate functions expects you to handle parallelism and streaming, but doesn't handle larger-than-memory state. I would say it should just be a scalar aggregate function if we were to implement it right now, though, since we haven't really started thinking about larger-than-memory state.

aocsa commented:
I updated this PR  exposing  only select_k_unstable using a optional parameter to choose strategy kind (Stable|NonStable).  Moreover python binding were added and their respectives tests. As we agreed In a follow up PR  a implementation for select_k_stable will be added.    cc @pitrou @lidavidm

cyb70289 commented on cpp/src/arrow/compute/kernels/vector_sort.cc:
```diff
@@ -1778,6 +1799,697 @@ class SortIndicesMetaFunction : public MetaFunction {
+      const auto rval = GetView::LogicalValue(arr.GetView(right));
+      return comparator(lval, rval);
+    };
+    arrow::internal::Heap<uint64_t, decltype(cmp)> heap(cmp);
+
+    uint64_t* iter = indices_begin;
+    for (; iter != kth_begin; ++iter) {
+      heap.Push(*iter);
+    }
+
```
Instead of implementing a heap by oursleves, can we use `std::priority_queue` directly here? Any special requirements a priority_queue cannot meet?

Another small catch is this heap initialization code is O(nlogn) complexity(insert elements one by one). `std::priority_queue` initializes heap directly from a vector, I suppose it's using O(n) approach like std::make_heap.

aocsa replied:
This internal_heap was used to implement a stable_heap. Stable algorithm for topk will be tackle in a follow-up PR so it's not needed for now, i'm removing this. Good catch with the `make_heap` approach. Thanks @cyb70289

lidavidm commented on cpp/src/arrow/compute/kernels/vector_sort.cc:
```diff
@@ -1778,6 +1799,711 @@ class SortIndicesMetaFunction : public MetaFunction {
+      : MetaFunction("top_k", Arity::Unary(), &top_k_doc, &kDefaultTopKOptions) {}
+
+  Result<Datum> ExecuteImpl(const std::vector<Datum>& args,
+                            const FunctionOptions* options,
+                            ExecContext* ctx) const override {
+    SelectKthMetaFunction impl;
+    const TopKOptions& opts = static_cast<const TopKOptions&>(*options);
+
+    std::vector<SortKey> sort_keys;
+    for (const auto& name : opts.keys)
+      sort_keys.emplace_back(SortKey(name, TopKOptions::order()));
+    if (args[0].kind() == Datum::ARRAY || args[0].kind() == Datum::CHUNKED_ARRAY) {
+      sort_keys.emplace_back(SortKey("not-used", TopKOptions::order()));
+    }
+    SelectKOptions select_k_options(opts.k, sort_keys, opts.kind);
+    return impl.ExecuteImpl(args, &select_k_options, ctx);
+  }
+};
+
+class BottomKMetaFunction : public MetaFunction {
+ public:
+  BottomKMetaFunction()
+      : MetaFunction("bottom_k", Arity::Unary(), &bottom_k_doc, &kDefaultBottomKOptions) {
+  }
+  Result<Datum> ExecuteImpl(const std::vector<Datum>& args,
+                            const FunctionOptions* options,
+                            ExecContext* ctx) const override {
+    SelectKthMetaFunction impl;
+    const BottomKOptions& opts = static_cast<const BottomKOptions&>(*options);
+
+    std::vector<SortKey> sort_keys;
+    for (const auto& name : opts.keys)
+      sort_keys.emplace_back(SortKey(name, BottomKOptions::order()));
+    if (args[0].kind() == Datum::ARRAY || args[0].kind() == Datum::CHUNKED_ARRAY) {
+      sort_keys.emplace_back(SortKey("not-used", BottomKOptions::order()));
+    }
+    SelectKOptions select_k_options(opts.k, sort_keys, opts.kind);
+    return impl.ExecuteImpl(args, &select_k_options, ctx);
+  }
+};
```
As stated above I would rather we not have explicit top_k and bottom_k functions if they're just going to dispatch to select_k. We should have only select_k, and top/bottom_k can be convenience constructors on the options themselves.

lidavidm commented on cpp/src/arrow/compute/kernels/vector_sort.cc:
```diff
@@ -1778,6 +1799,711 @@ class SortIndicesMetaFunction : public MetaFunction {
+                            const FunctionOptions* options, ExecContext* ctx) const {
+    const SelectKOptions& select_k_options = static_cast<const SelectKOptions&>(*options);
+    if (select_k_options.k < 0) {
+      return Status::Invalid("TopK/BottomK requires a valid `k` parameter");
```
```suggestion
      return Status::Invalid("SelectK requires a nonnegative `k`, got ", select_k_options.k);
```

lidavidm commented on cpp/src/arrow/compute/kernels/vector_sort.cc:
```diff
@@ -1778,6 +1799,711 @@ class SortIndicesMetaFunction : public MetaFunction {
+        break;
+    }
+    return Status::NotImplemented(
+        "Unsupported types for sort_indices operation: "
```
```suggestion
        "Unsupported types for select_k operation: "
```

lidavidm commented on cpp/src/arrow/compute/api_vector.cc:
```diff
@@ -140,6 +150,50 @@ PartitionNthOptions::PartitionNthOptions(int64_t pivot)
+      sort_keys(std::move(sort_keys)),
+      kind(kind) {}
+
+bool SelectKOptions::is_top_k() const {
+  SortOrder order = SortOrder::Descending;
+  for (const auto& k : sort_keys) {
+    order = k.order;
+    if (order != SortOrder::Descending) {
+      break;
+    }
+  }
+  return order == SortOrder::Descending;
+}
```
Why track `order`, instead of just iterating through the sort keys and returning `false` if the order is wrong, returning `true` outside the loop?

lidavidm replied:
```suggestion
bool SelectKOptions::is_top_k() const {
  for (const auto& k : sort_keys) {
    if (k.order != SortOrder::Descending) {
      return false;
    }
  }
  return true;
}
```

lidavidm commented on cpp/src/arrow/compute/kernels/vector_sort.cc:
```diff
@@ -1778,6 +1799,711 @@ class SortIndicesMetaFunction : public MetaFunction {
+  Result<Datum> ExecuteImpl(const std::vector<Datum>& args,
+                            const FunctionOptions* options,
+                            ExecContext* ctx) const override {
+    SelectKthMetaFunction impl;
```
Can we just inline the above class here? It doesn't seem useful to split them. (Once we remove TopKMetaFunction et al)

lidavidm commented on cpp/src/arrow/compute/kernels/vector_topk_benchmark.cc:
```diff
@@ -0,0 +1,65 @@
+namespace compute {
+constexpr auto kSeed = 0x0ff1ce;
+
+Result<std::shared_ptr<Array>> TopKWithSorting(const Array& values, int64_t n) {
```
This is unused?

aocsa replied:
ohh my bad, yes it is not used.

lidavidm reviewed (changes requested):
For CI: note that
- The new options classes need to be added to the tests: https://github.com/apache/arrow/pull/11019/checks?check_run_id=3540995813#step:8:7859
- Python code needs to be linted: https://github.com/apache/arrow/pull/11019/checks?check_run_id=3540995630#step:7:4148

lidavidm commented on cpp/src/arrow/compute/api_vector.h:
```diff
@@ -120,20 +120,29 @@ class ARROW_EXPORT SortOptions : public FunctionOptions {
+  static SelectKOptions TopKDefault(int64_t k, std::vector<std::string> key_names = {}) {
+    std::vector<SortKey> keys;
+    for (const auto& name : key_names)
+      keys.emplace_back(SortKey(name, SortOrder::Descending));
```
nit: please add braces for this sort of thing

aocsa replied:
ok thanks

lidavidm commented on cpp/src/arrow/compute/kernels/vector_sort.cc:
```diff
@@ -1826,7 +1823,7 @@ const FunctionDoc top_k_doc(
      "For floating-point types, NaNs are considered greater than any\n"
      "other non-null value, but smaller than null values."),
-    {"input"}, "TopKOptions");
+    {"input"}, "SelectKOptions::TopKDefault");
```
Note the param here should be just the name of the options class, i.e. "SelectKOptions". Python uses this for introspection.

lidavidm replied:
Though, why do we still even have top_k_doc and bottom_k_doc?

aocsa replied:
yeap it's old code

lidavidm commented on python/pyarrow/tests/test_compute.py:
```diff
@@ -127,11 +127,12 @@ def test_option_class_equality():
         pc.StrftimeOptions(),
     ]
-    classes = {type(option) for option in options}
+    classes={type(option) for option in options}
```
Why all these whitespace changes?

aocsa replied:
I noticed too, my editor tricked me

lidavidm commented on cpp/src/arrow/compute/api_vector.h:
```diff
@@ -131,8 +165,6 @@ class ARROW_EXPORT PartitionNthOptions : public FunctionOptions {
   int64_t pivot;
 };
 
-/// @}
```
Did you mean to remove this? This tells Doxygen where a group ends.

aocsa replied:
ohhh thanks I
 thought it was an error

lidavidm commented on cpp/src/arrow/compute/kernels/vector_sort.cc:
```diff
@@ -1778,6 +1798,650 @@ class SortIndicesMetaFunction : public MetaFunction {
+};
+
+template <typename ArrowContainer>
+static Status CheckConsistency(const ArrowContainer& container,
```
You can avoid templating this by taking `const Schema& schema` instead. You may also want to use `FieldRef::FindOne` since this will also tell you if the field can be unambiguously referenced. https://github.com/apache/arrow/blob/bb1ef850d3a6f5b998ff4dc1f196ee9cb7c273a3/cpp/src/arrow/type.h#L1631

lidavidm commented on cpp/src/arrow/compute/kernels/select_k_test.cc:
```diff
@@ -0,0 +1,774 @@
+};
+
+template <SortOrder order>
+Result<std::shared_ptr<Array>> SelectK(const Array& values, int64_t k) {
```
Do we need all these overloads? Datum has implicit constructors: https://github.com/apache/arrow/blob/bb1ef850d3a6f5b998ff4dc1f196ee9cb7c273a3/cpp/src/arrow/datum.h

lidavidm commented on cpp/src/arrow/compute/kernels/select_k_test.cc:
```diff
@@ -0,0 +1,774 @@
+  }
+}
+
+void ValidateSelectK(const ChunkedArray& chunked_array, int64_t k,
```
Same goes here - I would expect we only need one overload.

aocsa commented:
Thanks @lidavidm I updated this PR addressing last feedback comments.  cross-fingers, hope this time will be OK for merging.

lidavidm commented on docs/source/cpp/compute.rst:
```diff
@@ -1367,6 +1367,10 @@ value, but smaller than nulls.
+| select_k              | Unary      | Binary- and String-like     | UInt64            | :struct:`SelectKOptions`       | \(5) \(3) \(6) |
++-----------------------+------------+-----------------------------+-------------------+--------------------------------+----------------+
+| select_k              | Unary      | Boolean, Numeric, Temporal  | UInt64            | :struct:`SelectKOptions`       | \(5) \(6)      |
++-----------------------+------------+-----------------------------+-------------------+--------------------------------+----------------+
```
Why is this row duplicated twice? We can keep all the notes together (it should be clear that (3) only applies to binary/string types only…)

lidavidm replied:
Also, please try to keep the table sorted alphabetically.

lidavidm commented on docs/source/python/api/compute.rst:
```diff
@@ -350,6 +350,7 @@ Sorts and partitions
 
    partition_nth_indices
    sort_indices
+   select_k
```
Please try to keep this list sorted alphabetically.

lidavidm commented on cpp/src/arrow/compute/kernels/vector_sort.cc:
```diff
@@ -1778,6 +1798,627 @@ class SortIndicesMetaFunction : public MetaFunction {
+     "other non-null value, but smaller than null values."),
+    {"input"}, "SelectKOptions");
+
+Result<std::shared_ptr<ArrayData>> MakeMutableArrayForNumericBasedType(
+    std::shared_ptr<DataType> out_type, int64_t length, MemoryPool* memory_pool) {
+  auto buffer_size = BitUtil::BytesForBits(
+      length * std::static_pointer_cast<UInt64Type>(out_type)->bit_width());
+  std::vector<std::shared_ptr<Buffer>> buffers(2);
+  ARROW_ASSIGN_OR_RAISE(buffers[1], AllocateResizableBuffer(buffer_size, memory_pool));
+  auto out = std::make_shared<ArrayData>(out_type, length, buffers, 0);
+  return out;
+}
```
A few things here. Why accept out_type when it casts it to UInt64Type? Maybe just change this to `MakeMutableUInt64Array` or similar and use `sizeof(uint64_t)`.

lidavidm replied:
```suggestion
Result<std::shared_ptr<ArrayData>> MakeUInt64Array(MemoryPool* memory_pool, int64_t length) {
  auto buffer_size = length * sizeof(uint64_t);
  ARROW_ASSIGN_OR_RAISE(auto data, AllocateBuffer(buffer_size, memory_pool));
  return ArrayData::Make(uint64(), length, {nullptr, std::move(data)}, /*null_count=*/0);
}
```

lidavidm commented on cpp/src/arrow/compute/kernels/vector_sort.cc:
```diff
@@ -1778,6 +1784,711 @@ class SortIndicesMetaFunction : public MetaFunction {
+
+#undef VISIT
+
+  template <typename InType>
```
I suppose looking at this again we don't implement sorting for things like timestamp type in general(?) - we may want to file a followup to expand type support while trying to reduce the generated code.

aocsa replied:
I tried using VisitTypeInline and then create template specialization  for these  selecters however I got some error in the CI which I was not able to replicate at local. I will create an issue on JIRA for this case as well .

aocsa replied:
https://issues.apache.org/jira/browse/ARROW-13971

lidavidm replied:
Ah ok, thanks for checking!

lidavidm commented on cpp/src/arrow/compute/kernels/vector_sort.cc:
```diff
@@ -1778,6 +1798,627 @@ class SortIndicesMetaFunction : public MetaFunction {
+    StablePartitioner partitioner;
+    auto nulls_begin =
+        partitioner(indices_begin, indices_end, [&first_sort_key](uint64_t index) {
+          const auto chunk = first_sort_key.GetChunk<ArrayType>((int64_t)index);
```
Please use static_cast, not a C-style cast.

lidavidm commented on cpp/src/arrow/compute/kernels/select_k_test.cc:
```diff
@@ -0,0 +1,737 @@
+  ASSERT_OK_AND_ASSIGN(auto sorted_indices,
+                       SortIndices(datum, SortOptions({SortKey("unused", order)})));
+
+  if (k < datum.length()) {
```
Doesn't this mean we aren't asserting anything if k > datum.length()?

aocsa replied:
this is a corner case, when k > datum.length() `k =  datum.length()` if there is not nulls. If there are nulls, then k is located at the beginning of these nulls after partition. So `k =  select_k_indices.length()`;

lidavidm reviewed (commented):
Thanks for the updates, I think this is mostly shaping up.

aocsa commented:
Thanks again for the comments @lidavidm. Besides the minor latest changes addressing the feedback I filed follow-up JIRA issues: 

[C++][Compute] Improve top_k/bottom_k Selecters via CRTP
https://issues.apache.org/jira/browse/ARROW-13971

[C++][Compute]Implement SelectKStable
https://issues.apache.org/jira/browse/ARROW-13969

[C++][Compute]Implement streaming  version for SelectK
https://issues.apache.org/jira/browse/ARROW-13970

The first one is an improvement of the selecter and template specialization for specific type.  The last two are tasks to implement a streaming version of it and SelectKStable. 

Apart from that, I think it is ready for final review / merging, let me know what you think.

lidavidm reviewed (approved):
Thanks for addressing everything. I think we can merge, I'll also filed a followup to add a SelectKSinkNode: ARROW-13973

lidavidm commented:
@edponce or @cyb70289 any other comments?

cyb70289 commented:
@lidavidm, I don't have comments. Thanks @aocsa !

edponce commented on cpp/src/arrow/compute/api_vector.cc:
```diff
@@ -140,6 +143,29 @@ PartitionNthOptions::PartitionNthOptions(int64_t pivot)
+  for (const auto& k : sort_keys) {
+    if (k.order != SortOrder::Descending) {
+      return false;
+    }
```
Not major important, but I think there is a logic discrepancy for identifying topK/bottomK in `is_top_x` functions and how it is reasoned about it in the core SelectK code. Based on these loops, it is topK iff all keys are in descending order and it is bottomK iff all keys are in ascending order. So if keys order are mixed then it is neither, but in the SelectK logic ([here](https://github.com/apache/arrow/pull/11019/files#diff-bf5bdbbf3ffa5d706a841eb38face9739c18d515eeb829e31749c7deff2e9384R2347-R2352) and [here](https://github.com/apache/arrow/pull/11019/files#diff-bf5bdbbf3ffa5d706a841eb38face9739c18d515eeb829e31749c7deff2e9384R2060-R2063)) if is not one then it is assumed to be the other. Also, the visit code only looks at the first key.

This discrepancy only has implications if `SortOrder` is not a binary enum (which I do not foresee happening).
Nevertheless, for consistency, I recommend to make these loops identify "topness" based solely on the first key.
```c++
bool SelectKOptions::is_top_k() const {
  for (const auto& key : sort_keys) {
    return key.order == SortOrder::Descending;
  }
  return true;
}
```
Also, this bypasses the need to iterate through all the keys.

aocsa replied:
Agree! I will note this in the SelectKStable jira issue.

lidavidm commented:
Thanks @aocsa we can tackle some of these in followups.

edponce commented on cpp/src/arrow/compute/kernels/vector_sort.cc:
```diff
@@ -1778,6 +1798,621 @@ class SortIndicesMetaFunction : public MetaFunction {
+};
+
+template <SortOrder sort_order>
+class ArraySelecter : public TypeVisitor {
```
I am a bit queazy about the `SortOrder` template parameter. It is only used to select the corresponding [`SelectKComparator`](https://github.com/apache/arrow/pull/11019/files#diff-bf5bdbbf3ffa5d706a841eb38face9739c18d515eeb829e31749c7deff2e9384R1890) and it allows setting it to a different order than what is dictated by the `sort_keys` in `SelectKOptions`. To prevent this discrepancy, you could have the `SelectKComparator` be a private member of the `XXXSelecter` classes and set it based on `is_top_k()` etc. This would move the checks from [here](https://github.com/apache/arrow/pull/11019/files#diff-bf5bdbbf3ffa5d706a841eb38face9739c18d515eeb829e31749c7deff2e9384R2347-R2368) to the constructors.

aocsa replied:
Nice catch @edponce

lidavidm commented:
Sorry Eduardo, I had thought that comment was your review. Thanks for filling ARROW-13974, we can fix up things there.

aocsa commented on cpp/src/arrow/compute/api_vector.h:
```diff
@@ -252,6 +292,21 @@ ARROW_EXPORT
+/// \return a datum with the same schema as the input
+ARROW_EXPORT
+Result<std::shared_ptr<Array>> SelectKUnstable(const Datum& datum, SelectKOptions options,
+                                               ExecContext* ctx = NULLPTR);
```
In this case I think this API is similar to SortIndicesAPI

```Result<std::shared_ptr<Array>> SortIndices(const Array& array, SortOrder order = SortOrder::Ascending, ExecContext* ctx = NULLPTR);```

It should be clear someway  that with Default you will get and Status::Invalid error code.

edponce replied:
Well, ok I guess. I am curious why a `Status::Invalid` is returned if using the default options?

aocsa replied:
because `k` is not specified.

aocsa commented on cpp/src/arrow/compute/kernels/select_k_test.cc:
```diff
@@ -0,0 +1,736 @@
+
+template <typename ArrayType, SortOrder order>
+class SelectKComparator {
+ public:
```
I will rename this class as SelectKCompareForResult, the porpuse of this comparator is different. This is used to compare results, which include duplicates, not to use as a comparator to get the new order of data.

aocsa commented on cpp/src/arrow/compute/kernels/select_k_test.cc:
```diff
@@ -0,0 +1,736 @@
+    return SelectKUnstable(Datum(values), options);
+  } else {
+    return SelectKUnstable(Datum(values), options);
+  }
```
You are right, with simplication of SelectK API now the call to this function is direct.
